x/milestone/abci/abci.go

```diff
@@ -0,0 +1,240 @@
+	if err != nil {
+		return nil, err
+	}
+
+	blocksSinceLastMilestone := reqBlock - lastMilestoneBlockNumber
+
+	logger.Debug("blocksSinceLastMilestone", "blocksSinceLastMilestone", blocksSinceLastMilestone)
+
+	// TODO: make blocksSinceLastMilestone limit configurable
+	propStartBlock := uint64(0)
+	if pendingMilestone != nil && milestone != nil && blocksSinceLastMilestone > 6 {
+		propStartBlock = milestone.EndBlock + 1
+	} else {
+		if pendingMilestone != nil {
+			propStartBlock = pendingMilestone.StartBlockNumber + uint64(len(pendingMilestone.BlockHashes))
+		} else if milestone != nil {
+			propStartBlock = milestone.EndBlock + 1
+		} else {
+			propStartBlock = 0
+		}
+	}
+
+	blockHashes, err := getBlockHashes(ctx, propStartBlock, contractCaller)
+	if err != nil {
+		return nil, err
+	}
+
+	milestoneProp := &sidetxs.MilestoneProposition{
+		BlockHashes:      blockHashes,
+		StartBlockNumber: propStartBlock,
+	}
+
+	SetPendingMilestoneProposition(milestoneProp)
+
+	return milestoneProp, nil
+}
+
+func GetMajorityMilestoneProposition(ctx sdk.Context, validatorSet stakeTypes.ValidatorSet, extVoteInfo []abciTypes.ExtendedVoteInfo, logger log.Logger) (*sidetxs.MilestoneProposition, []byte, string, error) {
+	ac := address.HexCodec{}
+
+	hashToProp := make(map[string]*sidetxs.MilestoneProposition)
+	hashToVotingPower := make(map[string]int64)
+	hashToAggregatedProposersHash := make(map[string][]byte)
+	hashVoters := make(map[string][]string)
+	valAddressToVotingPower := make(map[string]int64)
+	totalVotingPower := validatorSet.GetTotalVotingPower()
+
+	for _, vote := range extVoteInfo {
+		// if not BlockIDFlagCommit, skip that vote, as it doesn't have relevant information
+		if vote.BlockIdFlag != cmtTypes.BlockIDFlagCommit {
+			continue
+		}
+
+		voteExtension := new(sidetxs.VoteExtension)
+		if err := voteExtension.Unmarshal(vote.VoteExtension); err != nil {
+			return nil, nil, "", fmt.Errorf("error while unmarshalling vote extension: %w", err)
+		}
+
+		if voteExtension.MilestoneProposition == nil {
+			continue
+		}
+
+		valAddr, err := ac.BytesToString(vote.Validator.Address)
+		if err != nil {
+			return nil, nil, "", err
+		}
+
+		_, validator := validatorSet.GetByAddress(valAddr)
+		if validator == nil {
+			return nil, nil, "", fmt.Errorf("failed to get validator %s", valAddr)
+		}
+
+		valAddressToVotingPower[valAddr] = validator.VotingPower
+
+		blockHashesCount := uint64(len(voteExtension.MilestoneProposition.BlockHashes))
+		prefix := make([][]byte, 0)
+		for i := uint64(0); i < blockHashesCount; i++ {
+			prefix = append(prefix, voteExtension.MilestoneProposition.BlockHashes[i])
+
+			hash := common.BytesToHash(bytes.Join(prefix, []byte{'|'})).String()
+			prefixCopy := make([][]byte, len(prefix))
+			copy(prefixCopy, prefix)
+			hashToProp[hash] = &sidetxs.MilestoneProposition{
+				BlockHashes:      prefixCopy,
+				StartBlockNumber: voteExtension.MilestoneProposition.StartBlockNumber,
```

Comment:
Can a malicious vote contain the same set of block hashes as the majority but a different `StartBlockNumber`  ?

Reply:
Thats good catch. We should include the `StartBlockNumber` in the `hash := common.BytesToHash(bytes.Join(prefix, []byte{'|'})).String()` because otherwise we can get a case where sometimes we pick the milestone proposition by the malicious proposer and don't finalize, because his start number is wrong.

Reply:
Fixed it.